### PR TITLE
Storage Driver for Aliyun OSS

### PR DESCRIPTION
**Pull request by denverdino**
This driver implements the storagedriver.StorageDriver interface and uses Aliyun OSS Open Storage Service as storage backend.

The documentation for this driver is in [https://github.com/denverdino/distribution/blob/master/docs/storage-drivers/oss.md](https://github.com/denverdino/distribution/blob/master/docs/storage-drivers/oss.md)

To run the unit test case for this driver, you need set the following environment variables before

```
export ALIYUN_ACCESS_KEY_ID=<YOUR_ACCESS_KEY_ID>
export ALIYUN_ACCESS_KEY_SECRET=<YOUR_ACCESS_SECRET>
export OSS_REGION=<OSS_REGION_ID>   E.g. oss-us-west-1
export OSS_BUCKET=<OSS_BUCKET_NAME>
export OSS_ENCRYPT=false
```

NOTE: This driver reuses the code logic from the S3 driver. 

Signed-off-by: Li Yi (denverdino@gmail.com) Boshi Lian(farmer1992@gmail.com)

### CODE REVIEW

File: registry/storage/driver/oss/oss.go [hunk 1]
```diff
@@ -0,0 +1,813 @@
...
+	if !ok || fmt.Sprint(regionName) == "" {
+		return nil, fmt.Errorf("No region parameter provided")
+	}
+
+	bucket, ok := parameters["bucket"]
+	if !ok || fmt.Sprint(bucket) == "" {
+		return nil, fmt.Errorf("No bucket parameter provided")
+	}
+
+	internalBool := false
```
Comment (noxiouz) [inline registry/storage/driver/oss/oss.go]: Why couldn't we use a scoped value inside if?
```go
internal, ok := parameters["internal"]
if ok {
		internalBool, ok := internal.(bool)
		if !ok {
			return nil, fmt.Errorf("The encrypt parameter should be a boolean")
		}
}
```

**Comment by stevvooe** [comment]
@denverdino This looks like a duplicate of #513. Which one would you like to keep?

**Comment by denverdino** [comment]
@stevvooe, Steve
>> This is looking like a straight copy paste from the s3 driver? Any way we can modify the s3 driver to support Aliyun OSS?

If we can abstract S3 driver with the interface, we can have the Aliyun OSS implementation. And currently, I provide the Aliyun OSS GO API in [https://github.com/denverdino/aliyungo](https://github.com/denverdino/aliyungo) which refers the API of [https://github.com/AdRoll/goamz](https://github.com/AdRoll/goamz). So, it should be easy to provide the OSS impl with that.

**Comment by stevvooe** [comment]
@denverdino Why can't the s3 driver be used directly? What is missing? What is different?

**Comment by denverdino** [comment]
@stevvooe , 
The current S3 driver is using the impl of github.com/AdRoll/goamz.  OSS API is similar to S3, but there are still  some differences, e.g. Authentication, Region definition, host for Bucket to access, etc. That is why, we cannot use the S3 driver directly.

So if we can have one abstract S3 driver, I can make the OSS Go API provide the compatibility layer to hide the implementation differences. 

Thanks

**Comment by stevvooe** [comment]
@denverdino Let us know if you'll continue work on this PR. We are closing for now.

**Comment by denverdino** [comment]
@stevvooe 
We will continue maintain the OSS driver. and we have already used it in several internal projects in Aliyun of Alibaba inc. It is pretty solid now.

Can you reopen this PR?

Thanks

**Comment by stevvooe** [comment]
@denverdino Reopened. Please rebase or merge, since a large number of commits seem to have diverged.

**Comment by denverdino** [comment]
@Stevooe

Thank you. I will update that.

File: registry/storage/driver/oss/oss.go [hunk 2]
```diff
@@ -0,0 +1,827 @@
...
+//
+// This package leverages the denverdino/aliyungo client library for interfacing with
+// oss.
+//
+// Because OSS is a key, value store the Stat call does not support last modification
+// time for directories (directories are an abstraction for key, value stores)
+//
+// Keep in mind that OSS guarantees only eventual consistency, so do not assume
+// that a successful write will mean immediate access to the data written (although
+// in most regions a new object put has guaranteed read after write). The only true
```
Comment (aaronlehmann) [inline registry/storage/driver/oss/oss.go]: This note about consistency looks like it's copied from the S3 driver. Is it really valid for OSS as well?

**Comment by dmp42** [comment]
@denverdino we fixed the build on master - can you rebase and push -f ?

Thanks!

File: registry/storage/driver/oss/oss.go (hunk 2 above)
Comment (denverdino) [inline registry/storage/driver/oss/oss.go]: Yes, OSS provide the similar capability to S3. Thanks

**Comment by denverdino** [comment]
I have rebased the PR. Thanks!

File: registry/storage/driver/oss/oss.go (hunk 2 above)
Comment (aaronlehmann) [inline registry/storage/driver/oss/oss.go]: AWS has different consistency guarantees in different regions. I was wondering if OSS also has differences between the regions. If not (or if we're not sure), we should remove the part that says "(although in most regions a new object put has guaranteed read after write)". It's a very minor thing - I just want to make sure the comment is accurate.

**Comment by denverdino** [comment]
Update the comments of the consistence model per Aaron's comment. 
Thanks

**Comment by aaronlehmann** [comment]
@denverdino: Sorry to request one more change... could you please add build tags to the new files so the driver can be enabled or disabled at compile time? See the rados driver for an example.

Once this is done, we should be able to merge this PR.

Thanks for the contribution!

**Comment by tg123** [comment]
Hi build tag has been added

should I change circle.yml and add `include_oss` into `DOCKER_BUILDTAGS`

**Comment by RichardScothern** [comment]
LGTM.  Thanks @tg123

File: docs/storage-drivers/oss.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
+<!--GITHUB
+page_title: Aliyun OSS storage driver
+page_description: Explains how to use the Aliyun OSS storage drivers
+page_keywords: registry, service, driver, images, storage, OSS, aliyun
+IGNORES-->
+
+# Aliyun OSS storage driver
+
+An implementation of the `storagedriver.StorageDriver` interface which uses Aliyun OSS for object storage.
```
Comment (stevvooe) [inline docs/storage-drivers/oss.md]: Please add a link to documentation.

File: registry/storage/driver/oss/oss.go [hunk 4]
```diff
@@ -0,0 +1,824 @@
...
+
+	// Validate that the given credentials have at least read permissions in the
+	// given bucket scope.
+	if _, err := bucket.List(strings.TrimRight(params.RootDirectory, "/"), "", "", 1); err != nil {
+		return nil, err
+	}
+
+	// TODO Currently multipart uploads have no timestamps, so this would be unwise
+	// if you initiated a new OSS client while another one is running on the same bucket.
+	// multis, _, err := bucket.ListMulti("", "")
```
Comment (stevvooe) [inline registry/storage/driver/oss/oss.go]: Please remove commented code.

File: registry/storage/driver/oss/oss.go [hunk 5]
```diff
@@ -0,0 +1,824 @@
...
+	client.SetEndpoint(params.Endpoint)
+	bucket := client.Bucket(params.Bucket)
+
+	// Validate that the given credentials have at least read permissions in the
+	// given bucket scope.
+	if _, err := bucket.List(strings.TrimRight(params.RootDirectory, "/"), "", "", 1); err != nil {
+		return nil, err
+	}
+
+	// TODO Currently multipart uploads have no timestamps, so this would be unwise
```
Comment (stevvooe) [inline registry/storage/driver/oss/oss.go]: TODO format: `TODO(you): ...`

File: registry/storage/driver/oss/oss_test.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
+// +build include_oss
+
+package oss
+
+import (
+	alioss "github.com/denverdino/aliyungo/oss"
```
Comment (stevvooe) [inline registry/storage/driver/oss/oss_test.go]: Please gofmt and goimports these imports.

File: registry/storage/driver/oss/oss_test.go [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+	"github.com/docker/distribution/context"
+	storagedriver "github.com/docker/distribution/registry/storage/driver"
+	"github.com/docker/distribution/registry/storage/driver/testsuites"
+	"io/ioutil"
+	//"log"
```
Comment (stevvooe) [inline registry/storage/driver/oss/oss_test.go]: Please remove.

**Comment by stevvooe** [comment]
Let's clear up this last set of nits and we should be good to go.

**Comment by tg123** [comment]
@stevvooe 
thanks 

1. fix with goimports
1. link added
1. todo formatted
1. comment codes removed
1. rebase with latest master

**Comment by stevvooe** [comment]
@tg123 Is there a link to documentation?

**Comment by tg123** [comment]
@stevvooe 
do you mean add a link to `storage-drivers/oss.md` in `docs/storagedrivers.md`

here is the diff 3eaab7d

**Comment by stevvooe** [comment]
@tg123 Imagine I am a user who has never heard of aliyun oss. Where can I go to get information about how it works?

**Comment by tg123** [comment]
hi @stevvooe 
product link was added in 3eaab7d

`http://www.aliyun.com/product/oss` is the link to the product

detail
`+- [oss](storage-drivers/oss.md): A driver storing objects in [Aliyun OSS](http://www.aliyun.com/product/oss).`


or anywhere else I need to add the link to ?

**Comment by stevvooe** [comment]
@tg123 I requested a link in `docs/storage-drivers/oss.md`, per https://github.com/docker/distribution/pull/514#discussion_r35935525.

This ensures the user doesn't have to click back to get further information.

**Comment by tg123** [comment]
@stevvooe 
Hi 
link added

but tests did not pass, any way to trigger them again?

**Comment by aaronlehmann** [comment]
I reran the tests.

**Comment by stevvooe** [comment]
LGTM

**Comment by aaronlehmann** [comment]
Merged. Thanks for your contribution!

**Comment by denverdino** [comment]
@aaronlehmann @stevvooe Thanks for your support

**Comment by longwuyuan** [comment]
Hi, please advise if this is wrong place to ask this. I intend to use Docker Containers running on Aliyun ECS Servers. Aliyun only has stock CentOS with a 2.6 kernel. So I intend to update kernel and run containers with docker-compose. Does the driver here mean that I can use a specific storage-driver with docker on the servers I described above and relocate /var/lib/docker on OSS ?

**Comment by aaronlehmann** [comment]
No, this storage driver is for the Docker Registry, which is a server that provides container images for `docker pull`. It's not involved in running containers.

**Comment by denverdino** [comment]
@longwuyuan, you can contact to me for this question with email denverdino@gmail.com.

And you can try the Docker Machine with ECS driver for using Docker on Aliyun, the simple tutorial is enclosed

https://github.com/denverdino/aliyungo/wiki/%08%E9%98%BF%E9%87%8C%E4%BA%91Docker-Machine%E9%A9%B1%E5%8A%A8

**Pull request merged**